internal/runners/packages/packages.go

```diff
@@ -54,32 +52,63 @@ func executePackageOperation(pj *project.Project, out output.Outputer, authentic
 		return locale.WrapError(err, "package_ingredient_err", "Failed to resolve an ingredient named {{.V0}}.", name)
 	}
 
+	// Check if this is an addition or an update
+	if operation == model.OperationAdded {
+		req, err := model.GetRequirement(pj.CommitUUID(), ingredient.Namespace, name)
+		if err != nil {
+			return errs.Wrap(err, "Could not get requirement")
+		}
+		if req != nil {
+			operation = model.OperationUpdated
+		}
+	}
+
 	parentCommitID := pj.CommitUUID()
 	commitID, fail := model.CommitPackage(parentCommitID, operation, name, ingredient.Namespace, version)
 	if fail != nil {
 		return locale.WrapError(fail.ToError(), "err_package_"+string(operation))
 	}
 
-	if !isHeadless {
-		err := model.UpdateProjectBranchCommitByName(pj.Owner(), pj.Name(), commitID)
-		if err != nil {
-			return locale.WrapError(err, "err_package_"+string(operation))
-		}
+	revertCommit, err := model.GetRevertCommit(pj.CommitUUID(), commitID)
+	if err != nil {
+		return errs.Wrap(err, "Could not get revert commit to check if changes we indeed made")
 	}
 
-	err = updateRuntime(pj.Source().Path(), commitID, pj.Owner(), pj.Name(), runbits.NewRuntimeMessageHandler(out))
-	if err != nil {
-		if !failures.Matches(err, runtime.FailBuildInProgress) {
-			return locale.WrapError(err, "Could not update runtime environment. To manually update your environment run `state pull`.")
+	orderChanged := len(revertCommit.Changeset) > 0
+
+	logging.Debug("Order changed: %v", orderChanged)
+
+	// Update project references to the new commit, if changes were indeed made (otherwise we effectively drop the new commit)
+	if orderChanged {
+		if !isHeadless {
+			err := model.UpdateProjectBranchCommitByName(pj.Owner(), pj.Name(), commitID)
+			if err != nil {
+				return locale.WrapError(err, "err_package_"+string(operation))
+			}
 		}
-		out.Notice(locale.Tl("package_build_in_progress",
-			"A new build with your changes has been started remotely, please run `state pull` when the build has finished. You can track the build at https://{{.V0}}/{{.V1}}/{{.V2}}.",
-			constants.PlatformURL, pj.Owner(), pj.Name()))
-	} else {
-		// Only update commit ID if the runtime update worked
 		if fail := pj.Source().SetCommit(commitID.String(), isHeadless); fail != nil {
 			return fail.WithDescription("err_package_update_pjfile")
 		}
+	} else {
+		commitID = parentCommitID
+	}
+
+	// Create runtime
+	rt, err := runtime.NewRuntime(pj.Source().Path(), commitID, pj.Owner(), pj.Name(), runbits.NewRuntimeMessageHandler(out))
+	if err != nil {
+		return locale.WrapError(err, "err_packages_update_runtime_init", "Could not initialize runtime.")
+	}
+
+	if !orderChanged && rt.IsCachedRuntime() {
```

Comment:
Thanks for making this story much easier than it was a few days ago @mdrohmann ! :)